dev/set_matrix.py

```diff
@@ -0,0 +1,427 @@
+    Converts a version specifier (e.g. "< 3") to a tuple of (operator, version).
+
+    Examples
+    --------
+    >>> get_operator_and_version("< 3")
+    (<built-in function lt>, '3')
+    >>> get_operator_and_version("!= dev")
+    (<built-in function ne>, 'dev')
+    """
+    regexp = r"([<>=!]+)([\w\.]+)"
+    m = re.search(regexp, ver_spec.replace(" ", ""))
+
+    if m is None:
+        raise ValueError(
+            "Invalid value for `ver_spec`: '{}'. Must match this regular expression: '{}'".format(
+                ver_spec, regexp,
+            )
+        )
+
+    return str_to_operator(m.group(1)), m.group(2)
+
+
+def process_requirements(requirements, version=None):
+    """
+    Examples
+    --------
+    >>> process_requirements(None)
+    []
+    >>> process_requirements(["foo"])
+    ['foo']
+    >>> process_requirements({"== 0.1": ["foo"]}, "0.1")
+    ['foo']
+    >>> process_requirements({"< 0.2": ["foo"]}, "0.1")
+    ['foo']
+    >>> process_requirements({"> 0.1, != 0.2": ["foo"]}, "0.3")
+    ['foo']
+    >>> process_requirements({"== 0.1": ["foo"], "== 0.2": ["bar"]}, "0.2")
+    ['bar']
+    >>> process_requirements({"== dev": ["foo"]}, "0.1")
+    []
+    >>> process_requirements({"< dev": ["foo"]}, "0.1")
+    ['foo']
+    >>> process_requirements({"> 0.1": ["foo"]}, "dev")
+    ['foo']
+    >>> process_requirements({"== dev": ["foo"]}, "dev")
+    ['foo']
+    >>> process_requirements({"> 0.1, != dev": ["foo"]}, "dev")
+    []
+    """
+    if requirements is None:
+        return []
+
+    if isinstance(requirements, list):
+        return requirements
+
+    dev_numeric = "9999.9999.9999"
+
+    if version == "dev":
+        version = dev_numeric
+
+    if isinstance(requirements, dict):
+        for ver_spec, packages in requirements.items():
+            op_and_ver_pairs = map(get_operator_and_version, ver_spec.split(","))
+            match_all = all(
+                op(LooseVersion(version), LooseVersion(dev_numeric if ver == "dev" else ver))
+                for op, ver in op_and_ver_pairs
+            )
+            if match_all:
+                return packages
+        else:
+            return []
+
+    raise TypeError("Should not reach here")
+
+
+def remove_comments(s):
+    """
+    Examples
+    --------
+    >>> code = '''
+    ... # comment 1
+    ...  # comment 2
+    ... echo foo
+    ... '''
+    >>> remove_comments(code)
+    'echo foo'
+    """
+    return "\n".join(l for l in s.strip().split("\n") if not l.strip().startswith("#"))
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--diff-only",
+        action="store_true",
+        help="If specified, ignore unchanged flavors or configs.",
+    )
+    # make `--ref-config` and `--diff-files` configurable to make it easier to test this script.
+    parser.add_argument(
+        "--ref-config",
+        default="https://raw.githubusercontent.com/mlflow/mlflow/master/{}".format(CONFIG_PATH),
```

Comment:
Nit: It might be slightly easier to interpret this if it the naming conventions were `--ref-versions-yaml` and `VERSIONS_YAML_PATH` instead.

Reply:
Done :)